src/SwedbankPay.Sdk/SwedbankPayHttpClient.cs

```diff
@@ -0,0 +1,227 @@
+            {
+                response = await this.client.SendAsync(msg, HttpCompletionOption.ResponseHeadersRead).ConfigureAwait(false);
+            }
+            catch (HttpRequestException e)
+            {
+                throw new BadRequestException(e);
+            }
+            catch (TaskCanceledException te)
+            {
+                throw new ApiTimeOutException(te);
+            }
+
+            if (response.IsSuccessStatusCode)
+            {
+                var res = await response.Content.ReadAsStringAsync();
+                this.logger.LogInformation(res);
+                return JsonConvert.DeserializeObject<T>(res, JsonSerialization.JsonSerialization.Settings);
+            }
+
+            var responseMessage = await response.Content.ReadAsStringAsync();
+            this.logger.LogInformation(responseMessage);
+            ProblemsContainer problems;
+            if (!string.IsNullOrEmpty(responseMessage) && IsValidJson(responseMessage))
+                problems = JsonConvert.DeserializeObject<ProblemsContainer>(responseMessage);
+            else if (response.StatusCode == HttpStatusCode.NotFound)
+                problems = new ProblemsContainer("id", "Not found");
+            else
+                problems = new ProblemsContainer("Other", $"Response when calling SwedbankPay was: '{response.StatusCode}'");
+
+            var ex = onError(problems);
+            throw ex;
+        }
+
+
+        internal async Task<T> HttpRequest
+            <T>(HttpRequestMessage requestMessage, Func<ProblemsContainer, Exception> onError, object payload = null)
+            where T : new()
+        {
+            UpdateRequest(requestMessage, payload);
+
+            HttpResponseMessage response;
+            try
+            {
+                response = await this.client.SendAsync(requestMessage, HttpCompletionOption.ResponseHeadersRead).ConfigureAwait(false);
+            }
+            catch (HttpRequestException e)
+            {
+                throw new BadRequestException(e);
+            }
+            catch (TaskCanceledException te)
+            {
+                throw new ApiTimeOutException(te);
+            }
+
+            if (response.IsSuccessStatusCode)
+            {
+                var res = await response.Content.ReadAsStringAsync();
+                this.logger.LogInformation(res);
+                return JsonConvert.DeserializeObject<T>(res, JsonSerialization.JsonSerialization.Settings);
+            }
+
+            var responseMessage = await response.Content.ReadAsStringAsync();
+            this.logger.LogInformation(responseMessage);
+            ProblemsContainer problems;
+            if (!string.IsNullOrEmpty(responseMessage) && IsValidJson(responseMessage))
+                problems = JsonConvert.DeserializeObject<ProblemsContainer>(responseMessage);
+            else if (response.StatusCode == HttpStatusCode.NotFound)
+                problems = new ProblemsContainer("id", "Not found");
```

Comment:
How far do we want to abstract this? Why not just let the HTTP exception bubble to the top, including the common 404 Not Found?

Reply:
Here it's a successful response with no Exception? Just a "ProblemResponse"? 
And that's why OnError throws a exception instead.

Reply:
I don't think we should return a ProblemsContainer for anything that isn't an actual ProblemJson. By checking for Not Found and creating a ProblemsContainer like we're doing here we're throwing away valuable information that may reside in the HTTP response content, for example.